### PR TITLE
Allow ignore patterns in other config files.

### LINKED ISSUE #5596: Support `.sqlfluffignore` configuration in `pyproject.toml`
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

In the spirit of reducing the number of config files required, it would be great if the `sqlfluffignore` configuration could be supported in `pyproject.toml`.

### Use case

_No response_

### Dialect

`pyproject.toml` configuration

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### LINKED ISSUE #6036: Add `.sqlfluffignore` Configuration to `pyproject.toml`
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

Currently, `.sqlfluffignore` is used to specify files or directories that `sqlfluff` should skip over. However, it would be convenient to manage these ignore patterns directly within `pyproject.toml`, allowing all `sqlfluff` configurations to be centralized in a single file.

### Use case

Add support for a `[tool.sqlfluff.ignore]` section in `pyproject.toml` where users can specify the same ignore patterns that would typically go into a `.sqlfluffignore` file.

Example:

```toml
[tool.sqlfluff]
dialect = "ansi"
paths = ["./models"]

[tool.sqlfluff.ignore]
patterns = ["./venv", "./target"]
```

## Benefits

- **Centralized Configuration:** All sqlfluff settings are in one place.
- **Improved Workflow:** Avoids the need to manage separate .sqlfluffignore files.
- **Consistency:** Aligns with other tools using pyproject.toml for configuration.

I believe this enhancement would significantly improve sqlfluff usability and promote better project organization.

Thank you for considering this feature request.

### Dialect

All

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by alanmcruickshank**
This resolves #6036 and resolves #5596. This builds on the restructuring of the `config` module (see #6128 & #6177). This also uses a little of the code from #6206.

This allows ignore patterns to be defined in `.sqlfluff` and `pyproject.toml` files. For this to work, we effectively end up loading config files in two phases, once while processing any ignore patterns and then later when sequencing files for parsing. That's why it's so important for the caching to be in place so we don't just duplicate a bunch of filesystem operations.

Under the hood, this is basically a re-write for the `discovery` module, but I think it's much better for it even without the new functionality.

### CODE REVIEW

File: src/sqlfluff/core/linter/discovery.py [hunk 1]
```diff
@@ -54,13 +50,72 @@ def _check_ignore_specs(
...
+    # Use normalised path to ensure reliable caching.
+    config_dict = load_config_file_as_dict(Path(filepath).resolve())
+    ignore_section = config_dict.get("core", {})
+    if not isinstance(ignore_section, dict):
+        return None  # pragma: no cover
+    patterns = ignore_section.get("ignore_paths", [])
+    if isinstance(patterns, str):
+        patterns = patterns.split(",")
+    elif not patterns or not isinstance(patterns, list):
+        return None
```
Comment (WittierDinosaur) [inline src/sqlfluff/core/linter/discovery.py:96]: this probably wants an `else` to catch edge cases and throw an exception

Comment (alanmcruickshank) [reply src/sqlfluff/core/linter/discovery.py:96]: No need in this case, not being caught by either of these clauses means it's already a list and so doesn't need any edits. I'll add a comment to that explain that though 👍

**Review by WittierDinosaur — APPROVED** [review]
LGTM

**Review by keraion — APPROVED** [review]
LGTM

**Pull request merged**